flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -91,6 +91,21 @@ class ExprLowering {
     return builder.create<fir::ShapeShiftOp>(loc, shapeType, shapeArgs);
   }
 
+  fir::BoxAddressValue genBoxAddress(const Fortran::lower::SomeExpr &expr) {
+    // TODO: GetLastSymbol is not the right thing to do if expr if an
+    // allocatable or pointer derived type component.
+    auto *sym = Fortran::evaluate::GetLastSymbol(expr);
+    if (!sym)
+      mlir::emitError(getLoc(), "trying to get descriptor address of an "
+                                "expression that is not a variable");
+    return symMap.lookupSymbol(*sym).match(
+        [&](const Fortran::lower::SymbolBox::PointerOrAllocatable &boxAddr)
+            -> fir::BoxAddressValue { return boxAddr; },
```

Comment:
I'm sort of leery about always boxing all POINTERs and ALLOCATABLEs, as that was the original approach taken with CHARACTER, but it ultimately ended up being redone as it led to significant amounts of glue code doing nothing but packaging and unpackaging a LEN constant.
In this case, it seems very likely we're going to have small, local ALLOCATABLE records and sequences that don't require all the extra packaging and unpackaging as they'll just be a heap pointer and some constant(s) or trivial expressions based on arguments.

Reply:
The character boxing/unboxing was motivated by a technical limitation of lowering at the time: we were propagating things as bare `mlir::Value`. I agree it would have been much better to introduce the extended value right away.

The pointer/allocatable unboxing on reference is motivated by Fortran semantics. The boxed values may change. It is an optimization to assume they will not, and it requires some level of flow analysis to validate that we are in the simple case you are referring to. I think fir should be able to make the analysis and promote the box to simple heap if it is OK.

Now, I agree it will produce a bit of an unpleasant FIR where we have unboxing every time something is changed.
Note that dead code elimination will right away get rid of all the descriptor properties that we read but do not use.
That leaves all the redundant reads. If fir is able to prove the box has not changed between two read, it could merge the reads.

I can spent some time exploring what would be needed if we want to do that in lowering already, my feeling is that it will add a lot of complexity in lowering and we would probably misses some "promotion to simple heap" opportunities that we would have after fir inlining + folding pass.

If we want to keep the read-only aspect of the symbol table, that means we need to prove that the allocatable is allocated exactly one time in its lifetime, and that the ssa values defining its allocated properties at the point of its allocation are accessible at all use points. Basically the only programs where we could skip the box in lowering would look like:
```
! Type your code here, or load an example.
subroutine square(n)
    implicit none
    real, allocatable :: c(:)
    integer :: n
    ! No reference to c before this
    ! allocation at the top level so that ssa values can be used
    ! everywhere after that.
    allocate(c(n))
    ! No reallocation, explicit or implicit:
    !  - c not passed as input(out) or input(inout) allocatable
    !  - c not assigned in full array assignement c=
    !  - c not deallcoated + reallocated
    deallocate(c)
    ! no reference to c after that 
end subroutine
```

Otherwise if we want to avoid box creation case, we need to start mutating the symbol table according to what happened. In single thread, we may be able to manage this. In parallel contexts, that seems very hard to me. Is the risk/complexity in lowering worth the perf benefit of being able to avoid creating boxes in some cases ?

Example of a parallel openmp program where the allocation is happening in one thread and its used in others. The allocation is guaranteed to happen before all usages in the execution, but is syntactically after its usages by other threads:
```
module test
contains
subroutine square(n, x)
    use omp_lib
    real, allocatable :: x(:,:)
    integer :: n, id
    logical not_master
    integer(kind = OMP_lock_kind) :: lck
    call OMP_init_lock(lck)
    !$OMP PARALLEL SHARED(lck, x, n) PRIVATE(id)
    id = OMP_get_thread_num() + 1
    !$OMP MASTER
    call OMP_set_lock(lck)
    !$OMP END MASTER
    if (id.ne.1) then
      ! Wait for master
      call OMP_set_lock(lck)
        print *, "hi :", id
      call OMP_unset_lock(lck)
      ! Do something with x
      x(:, id) = id
    else
      if (.not. allocated(x)) then
        allocate(x(n, OMP_get_num_threads()))
      end if
      ! Let workers start working
      print *, "master"
      call OMP_unset_lock(lck)
      x(:, id) = id
    end if
    !$OMP END PARALLEL
end subroutine
end module
 use test
 real, allocatable :: x(:, :)
 call square(1, x)
 print *, x
end
```

I think it would also be possible to hide all the openmp constructs and to make it really hard for the compiler to now something fishy is going on, aside being compiled with -fopenmp.
Technically, I think `x` should have a volatile attribute but even without it, gfortran -fopenmp compiles this OK and running it prints [1, 2, 3, ...., num_therad].

Reply:
MutableBoxValue and Allocatable.h interface were reworked to remove the assumption that allocatable == descriptors.
For local variables were we know this is safe, working with local variables to keep track of allocatable properties was enabled. See updated test case `test/Lower/allocatables.f90`.
For other cases where there already is a descriptor, and enforcing a single point of truth for the properties is important (dummies, globals, volatiles), the original approach of directly working on the fir.ref<fir.box<>> is kept for now.
So MutableBoxValue and allocatable lowering implementation should not be be biased towards one solution or another.